src/redis-cli.c

```diff
@@ -802,7 +802,11 @@ static int cliSelect(void) {
     reply = redisCommand(context,"SELECT %d",config.dbnum);
     if (reply != NULL) {
         int result = REDIS_OK;
-        if (reply->type == REDIS_REPLY_ERROR) result = REDIS_ERR;
+        if (reply->type == REDIS_REPLY_ERROR) {
+            result = REDIS_ERR;
+            fprintf(stderr,"SELECT %d failed: %s\n",config.dbnum,reply->str);
+            config.dbnum = 0;
```

Comment:
it looks like this method is also run when the user randomly types `AUTH` in the interactive CLI (not sure why we do that), in which case, resetting to 0, might not be the right thing?
although considering that this `config.dbnum` should already be selected, i can't think of a reason why SELECT would fail.

can you please look into that?

Reply:
I saw this code too. I think the reason`SELECT` after `AUTH` is if we execute `redis-cli -n 2`, if user is not authed, `SELECT` command will not be sent. Then user executes `AUTH`, `SELECT` can be sent.
If we want to handler this, we should distinguish between select input and current select.

Reply:
i guess we need to distinguish between startup select and runtime.

or maybe have a separate variable for the db requested by command line, and the current selected db.
i.e. when a new connection is made, the initial db is always 0, and we only change it if we successfully executed a SELECT command (for whatever reason, either startup or later).